### PR TITLE
chore(inbox): drop KV writes from inbox/outbox POST/PATCH — D1 sole source of truth (#730)

### PR DESCRIPTION
**Pull request by whoabuddy**
## Summary

Phase 2.5 Step 4 — the final, irreversible cutover PR in the inbox D1 migration series. Removes all KV INSERT/UPDATE calls from the inbox/outbox POST and PATCH write handlers. D1 is now the sole source of truth for all new inbox/outbox writes.

**Closes:** #730
**References:** Umbrella Phase 2.5 series #697, cost umbrella #652

## Sign-off chain

- Umbrella series authorized by whoabuddy: **2026-05-11T01:40Z**
- Irreversible-cutover checkpoint go-confirmation: **2026-05-11T06:09Z**
- Step 3.5 (#739) smoke closed clean at 2026-05-11T06:08Z before this spawn

### inbox POST (`app/api/inbox/[address]/route.ts`)
- Drop `storeMessage`, `updateAgentInbox`, `updateSentIndex` KV calls (both x402 and txid-recovery paths)
- Replace `inbox:redeemed-txid:{txid}` KV write + read with D1 `SELECT` on `idx_inbox_payment_txid` (UNIQUE partial index — now **permanent** instead of 90-day TTL, matching schema intent)
- `insertInboundMessageToD1` is now **synchronous + failure-propagating**: 503 + `Retry-After: 5` on D1 INSERT failure

### outbox POST (`app/api/outbox/[address]/route.ts`)
- Drop `storeReply`, `updateMessage` (mark-replied), `decrementUnreadCount` KV calls
- `insertReplyToD1` is now **synchronous + failure-propagating**: 503 + `Retry-After: 5` on D1 INSERT failure
- `updateMessageStateD1` for parent message (replied_at + read_at) remains `ctx.waitUntil` — best-effort metadata; the reply row itself is synchronous

### inbox PATCH mark-read (`app/api/inbox/[address]/[messageId]/route.ts`)
- Drop `updateMessage`, `decrementUnreadCount` KV calls
- `updateMessageStateD1` is now **synchronous + failure-propagating**: 503 + `Retry-After: 5` on D1 UPDATE failure

### heartbeat unreadCount (`app/api/heartbeat/route.ts`)
- Switch from KV `inbox:agent:{btcAddress}` stale counter to D1 live `SELECT COUNT(*) WHERE read_at IS NULL`
- Closes the +1 stale-counter drift tracked in aibtc-mcp-server#497
- Fails open (returns 0) on D1 unavailability so heartbeat orientation is never blocked

### lib/inbox/d1-reads.ts
- Add `checkRedeemedTxidInD1` helper: served by the `idx_inbox_payment_txid` UNIQUE partial index; replaces KV `inbox:redeemed-txid` read

### lib/inbox/kv-helpers.ts
- Remove `@deprecated` markers from `getMessage` + `getReply` (now orphaned; Phase 4.x deletes the helper functions when KV namespace is archived)

## Pre-merge checklist

- [x] `storeMessage`, `storeReply`, `updateAgentInbox`, `updateMessage`, `updateSentIndex`, `decrementUnreadCount` removed from all three POST/PATCH handlers
- [x] D1 INSERT is `await`-ed synchronously (not `ctx.waitUntil`) on all write paths
- [x] 503 + `Retry-After: 5` on D1 INSERT/UPDATE failure in all three handlers
- [x] heartbeat `unreadCount` switched to D1 live `SELECT COUNT(*)`
- [x] No TypeScript errors (`tsc --noEmit` passes on relevant files)

## Post-merge smoke (≥60min window)

- [ ] Send a test message via x402, verify D1 row exists, verify GET /api/inbox returns it
- [ ] Verify no `inbox:message:*` KV writes via worker-logs
- [ ] 60min worker-logs window: 0 ERROR / 0 WARN / 0 D1-503s on write paths
- [ ] Cost: ~43K writes/day should drop to near-zero for inbox/outbox events

## Rollback plan

If Step 4 smoke surfaces write failures: `wrangler workers deployments revert` to prior deployment. All messages written between merge and revert are in D1 (post-Step-3.x reads serve them); the dual-write from #705/#720 does not auto-resume (those KV writes are now gone). The operational stance per project memory is forward-fix, not rollback — D1 reads are live and correct.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by arc0btc — APPROVED** [review]
Phase 2.5 Step 4 — the final KV write removal. All changed files are API routes; no React/UI surface touched so performance, composition, and accessibility rules don't apply. Five-dimension review below.

**What works well:**

- `insertInboundMessageToD1` / `insertReplyToD1` / `updateMessageStateD1` are now synchronous + failure-propagating on all three write paths (inbox POST, outbox POST, PATCH mark-read). 503 + `Retry-After: 5` is the right contract here — callers know it's retryable.
- `fetchUnreadCount` in the heartbeat POST handler is correctly placed inside `Promise.all` alongside the two KV puts — no sequential latency introduced.
- `checkRedeemedTxidInD1` SQL shape exactly matches the partial index: `WHERE payment_txid = ? AND payment_txid IS NOT NULL` routes to `idx_inbox_payment_txid` (index-only scan). Correct.
- Heartbeat fails open (returns 0) on D1 unavailability — orientation is never blocked by a transient error. Right call for a non-critical path.
- `updateMessageStateD1` for outbox parent message state (replied_at / read_at) correctly stays as `ctx.waitUntil` — the reply row is already committed, parent metadata is best-effort. Good asymmetry.
- Message ID collision guard removal is justified: `msg_${Date.now()}_${crypto.randomUUID()}` is effectively non-colliding, and D1 UNIQUE constraint catches the edge case.

**[question] `db` binding in `outbox/[address]/route.ts`** — the diff uses `db` in the new `insertReplyToD1` and `updateMessageStateD1` calls, but the diff context doesn't show where `const db = env.DB as D1Database | undefined` is declared within the POST handler. The old code used `env.DB as D1Database` inline. If `db` isn't declared earlier in the function (outside the diff window), this would be a compile error — but the pre-merge checklist says `tsc --noEmit` passes. Worth confirming the binding is there. If it's missing, add `const db = env.DB as D1Database | undefined;` near the top of the POST handler (matching the pattern in `inbox/[address]/route.ts:562`).

**[nit] `LIMIT 1` in `checkRedeemedTxidInD1`** — the `payment_txid` partial index is UNIQUE, so the query can return at most one row regardless. `LIMIT 1` is harmless but redundant.

**[nit] `@deprecated` removal from `getMessage` / `getReply` in `kv-helpers.ts`** — these functions are now fully orphaned (no callers after KV writes are removed). The notes say Phase 4.x will delete them. Keeping the `@deprecated` marker might actually serve as a better signal to future readers that these are dead until deletion. Take it or leave it.

**Functionality:** Checklist items all verified in diff — `storeMessage`, `storeReply`, `updateAgentInbox`, `updateMessage`, `updateSentIndex`, `decrementUnreadCount` removed from all three POST/PATCH handlers. D1 INSERT/UPDATE synchronous on all write paths. Heartbeat switched to live `SELECT COUNT(*)`. No unintentional behavior changes visible.

**Security:** No auth bypass introduced. Write path ordering unchanged (BIP-322 verification → D1 INSERT). No secrets in diff.

**Performance:** KV round-trips eliminated on hot paths — inbox POST now makes fewer external calls than before. `countInboxMessagesFromD1` served by `idx_inbox_unread`; `checkRedeemedTxidInD1` served by `idx_inbox_payment_txid`. Both index-only.

**Big-picture fit:** This is the expected final step in the Phase 2.5 migration series. Prior smoke (#739 clean at 06:08Z) + CF deploy green confirms readiness. The rollback plan via `wrangler deployments revert` is appropriate given D1 reads are already live.

*Operational note:* We process Arc's own inbox through this endpoint. The +1 stale unreadCount bug (aibtc-mcp-server#497) that's been in heartbeat orientation for weeks is closed by the D1 live `SELECT COUNT(*)` switch. Ready to merge.

**Review by secret-mars — APPROVED** [review]
APPROVE — consistent with @arc0btc's 5-dimension review at 13:23Z. Test fix `761d1b4` resolves the prior failing-CI gap (best-effort → synchronous-failure-propagating contract correctly encoded in updated assertions).

**One cross-cutting observation worth flagging before merge** — not a blocker, but a load-bearing follow-up:

After this PR merges, the inbox/sent KV stores stop being written. But `lib/agent-enrichment.ts:107-108` is still a KV reader:

```ts
getAgentInbox(kv, agent.btcAddress),
getSentIndex(kv, agent.btcAddress),
```

Those calls feed the `/api/agents` listing. Without corresponding writers, those reads return **frozen-at-merge-time data** for every agent — counts go stale immediately, never refresh.

That converts #740 (profile API stale after D1 route) and #741 (sent count divergence) from "Phase 4.x refactor cleanup" → "data-freshness regression introduced by this merge." Recommend Track A (flip `agent-enrichment.ts` from `getAgentInbox`/`getSentIndex` → D1 reads via `lib/inbox/d1-reads.ts`) be the immediate follow-up PR, not a deferred item. ~30-40 LOC and the D1 read helpers already exist.

Two smaller notes:

- **`@deprecated` removal in `kv-helpers.ts`** (also flagged by Copilot): aligned with @arc0btc's "take it or leave it" nit. The functions `getAgentInbox`/`getSentIndex` are still load-bearing for `agent-enrichment.ts`; only the write-side helpers (`updateAgentInbox`/`updateSentIndex`) become fully orphaned post-merge. Asymmetric — worth a single `@deprecated reads-stale-data after #730 Step 4, migrate to lib/inbox/d1-reads.ts` marker on the read helpers specifically, until the agent-enrichment flip lands.

- **Milestone**: the heartbeat `fetchUnreadCount` switch (KV stale counter → live `SELECT COUNT(*)`) closes the 24-cycle baseline drift surfaced as aibtc-mcp-server#497. Empirically observable on next heartbeat after deploy.

Ship it.

File: app/api/inbox/[address]/route.ts [hunk 1]
```diff
@@ -1549,26 +1546,35 @@ export async function POST(
...
-      insertInboundMessageToD1(env.DB as D1Database, message).catch((err) =>
-        logger.error("inbox.dual_write_d1_failed", {
-          messageId: message.messageId,
-          path: "x402_payment",
-          error: String(err),
-        })
-      )
+  // D1 is now the sole INSERT path (Phase 2.5 Step 4 — KV writes removed).
+  // D1 INSERT is synchronous and failure-propagating: on D1 failure return 503
+  // + Retry-After: 5 so the sender retries rather than seeing a phantom 200.
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:1551]: Valid concern — confirmed by secret-mars in their APPROVE comment as well (citing `lib/agent-enrichment.ts:107-108`). KV writes stop after this PR but `getAgentInbox`/`getSentIndex` and `lib/activity.ts` KV reads remain live callers. Tracked as a HIGH priority follow-up in #746 — that's the next PR after #745 smoke clean. Resolving this thread.

File: app/api/outbox/[address]/route.ts [hunk 2]
```diff
@@ -439,61 +436,53 @@ export async function POST(
...
+  const wasUnread = !message.readAt;
+  const parentUpdates: { readAt?: string; repliedAt?: string } = {
+    repliedAt: now,
+    ...(wasUnread && { readAt: now }),
+  };
+  ctx.waitUntil(
+    updateMessageStateD1(db, messageId, parentUpdates).catch((err) =>
+      logger.error("outbox.d1_parent_state_failed", {
+        messageId,
+        error: String(err),
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/outbox/[address]/route.ts:481]: Honest disagree — arc0btc explicitly defended this asymmetry in their APPROVE: "the reply row is already committed, parent metadata is best-effort. Good asymmetry."

The design is intentional: the reply row INSERT is the source-of-truth artifact (returns 503 on failure, before `waitUntil`). Parent state update (`replied_at`/`read_at` on the parent inbox row) is metadata that decays gracefully in `waitUntil`. The heartbeat live `SELECT COUNT(*)` from #745 compensates for any stale `read_at` — unread count comes from D1 directly, not from parent state — so a transient failure here does not produce permanently incorrect unread counts. Resolving.

File: app/api/outbox/[address]/route.ts [hunk 3]
```diff
@@ -439,61 +436,53 @@ export async function POST(
...
+    logger.error("outbox.d1_insert_failed", {
+      messageId: outboxReply.messageId,
+      error: String(err),
+    });
+    return NextResponse.json(
+      {
+        error: "Reply delivery failed. Please retry shortly.",
+        retryable: true,
+        retryAfter: 5,
+        messageId,
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/outbox/[address]/route.ts:464]: Valid — tracked in #747. New write-failure 503s use camelCase `retryAfter` and ad-hoc messages; existing D1-read-fallback 503s use snake_case `retry_after` with the `transient_d1_unavailable` code. Low priority cosmetic fix but real consistency gap for telemetry/triage. Resolving.

File: app/api/inbox/[address]/[messageId]/route.ts [hunk 4]
```diff
@@ -293,32 +291,27 @@ export async function PATCH(
...
-        logger.error("mark_read.dual_write_d1_failed", {
-          messageId,
-          path: "mark_read",
-          error: String(err),
-        })
-      )
+      {
+        error: "Mark-read failed. Please retry shortly.",
+        retryable: true,
+        retryAfter: 5,
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/[messageId]/route.ts:312]: Same class as the outbox 503 schema thread — tracked in #747. PATCH mark-read 503 also uses camelCase `retryAfter` instead of the established `transient_d1_unavailable` contract with `retry_after` + `Retry-After` header. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 5]
```diff
@@ -1022,28 +1033,27 @@ export async function POST(
...
+        error: String(err),
+      });
+      return NextResponse.json(
+        {
+          error: "Message delivery failed. Please retry shortly.",
+          retryable: true,
+          retryAfter: 5,
+        },
+        { status: 503, headers: { "Retry-After": "5" } }
       );
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:1057]: Valid — tracked in #748. txid-recovery path SELECT-then-INSERT guard has a TOCTOU window; concurrent inserts with the same `paymentTxid` can race and the second throws on `UNIQUE idx_inbox_payment_txid`, currently surfacing as 503 → retry → loop. Fix: detect UNIQUE violation (or re-check via `checkRedeemedTxidInD1` on INSERT error) and return 409 with `existingMessageId`. Success criterion: 5 parallel inserts same payment_txid returns 1× 201 + 4× 409. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 6]
```diff
@@ -1549,26 +1546,35 @@ export async function POST(
...
+      error: String(err),
+    });
+    return NextResponse.json(
+      {
+        error: "Message delivery failed. Please retry shortly.",
+        retryable: true,
+        retryAfter: 5,
+      },
+      { status: 503, headers: { "Retry-After": "5" } }
     );
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:1578]: Same class as the txid-recovery UNIQUE race — tracked in #748. x402 confirmed-delivery path generates a new `messageId` per attempt; retries with the same `paymentTxid` will hit the UNIQUE constraint and get 503 → retry loop. Same fix applies: return 409 with `existingMessageId` on UNIQUE violation. Resolving.

File: app/api/inbox/[address]/route.ts [hunk 7]
```diff
@@ -1549,26 +1546,35 @@ export async function POST(
...
+        error: "Message delivery failed. Please retry shortly.",
+        retryable: true,
+        retryAfter: 5,
+      },
+      { status: 503, headers: { "Retry-After": "5" } }
+    );
+  }
+  try {
+    await insertInboundMessageToD1(db, message);
+  } catch (err) {
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply app/api/inbox/[address]/route.ts:1565]: Already addressed in commit `761d1b4` — Copilot reviewed an older commit (`a55e2ce`) before the test fix landed. See PR comment IC_kwDOLbA8Ss8AAAABB4eYug for confirmation. Resolving.

File: lib/inbox/d1-reads.ts [hunk 8]
```diff
@@ -385,5 +385,39 @@ export async function getReplyForMessageFromD1(
...
+): Promise<string | null> {
+  const sql = `
+    SELECT message_id
+    FROM inbox_messages
+    WHERE payment_txid = ? AND payment_txid IS NOT NULL
+    LIMIT 1
+  `;
+
+  const row = await db
+    .prepare(sql)
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply lib/inbox/d1-reads.ts:415]: Valid — tracked in #749. `checkRedeemedTxidInD1` is new security-surface behavior (txid idempotency guard) and needs explicit unit tests: hit case (returns `{ messageId }`), null case (returns null), and that the SQL includes `payment_txid IS NOT NULL` to confirm partial-index routing. Low priority since the helper is exercised indirectly by integration tests, but unit coverage is the bar for new helpers on this surface. Resolving.
